### PR TITLE
Support pyproject.toml

### LINKED ISSUE #80: sdist is missing tests
The sdist at PyPI is missing tests.  Namely, there are missing both `examplesetup.py` and `tests/test_exampleproj.py` files, but `tox.ini` is included in the sdist.  Please add missing files for testing support to the sdist.  Thank you.

### LINKED ISSUE #88: Dependency cycle with twisted
Hello,
I was checking several repositories maintaining [python-incremental](https://archlinux.org/packages/extra/any/python-incremental/) and [python-twisted](https://archlinux.org/packages/extra/any/python-twisted/) where `incremental` depends on `twisted`, and `twisted` depends on `incremental`.

I checked  [setup.py](https://github.com/twisted/incremental/blob/trunk/setup.py) but I don't see any thing saying that `incremental` depends on `twisted`.

So, is `twisted` a mandatory dependency on `incremental`?

### LINKED ISSUE #90: Provide a way to leverage incremental from pyproject.toml without importing the target package
From discussion on https://github.com/twisted/treq/pull/388 —

> I don't see anything here equivalent to the use_incremental=True argument to setup(). It looks like we need to add support to incremental for direct integration with setuptools, like setuptools_scm has: https://github.com/pypa/setuptools_scm/blob/102f56a896bc4ea8d4c58d9c49f455f15310de38/pyproject.toml#L76-L77

The way that we presently allow the ability to have incremental eval a module without a full import is only via the `distutils.setup_keywords` entrypoint, [which is deprecated and heavily discouraged](https://setuptools.pypa.io/en/latest/userguide/extension.html#adding-arguments) as it doesn't work with pyproject.toml.

In Treq, and potentially in other projects, importing the package itself loads in a bunch of the package's submodules, which then pulls in dependent packages, so to get Treq off of setup.py, we would need to have its build dependencies fully replicate its runtime dependencies, which is obviously not ideal.

### PR DESCRIPTION
**Pull request by twm**
Add support for the [`setuptools.finalize_distribution_options` entry point](https://setuptools.pypa.io/en/latest/userguide/extension.html#customizing-distribution-options) so that Incremental can be configured with `pyproject.toml` alone. This is patterned off of the interface [setuptools_scm documents](https://github.com/pypa/setuptools_scm?tab=readme-ov-file#pyprojecttoml-usage): the user creates a ``[tool.incremental]`` section in ``pyproject.toml`` 

The actual implementation is quite different from the distutils one. The `finalize_distribution_options` hook always seems to be passed an empty `Distribution` object which is unable to list the files in the package. I settled on extracting the project name from `pyproject.toml` and reusing the logic from the `incremental.update` module to locate `_version.py`. As a side effect I reworked that logic to avoid a Twisted dependency for `FilePath`.

Also:

- Tests now check that `__version__` and the package metadata match
- The project uses `pyproject.toml` instead of `setup.cfg`
- Drop a bunch of Python 2.7 compatibility code

Fixes #90. Fixes #88. Fixes #80.

### CODE REVIEW

File: .coveragerc [hunk 1]
```diff
@@ -4,6 +4,10 @@ source_pkgs = incremental
 # List of directory names.
 source = tests
 branch = True
+parallel = True
+# This must be an absolute path because the example tests
+# run Python processes with alternative working directories.
+data_file = ${TOX_INI_DIR-.}/.coverage
```
Comment (glyph) [inline .coveragerc:10]: NB: `coverage` also supports `pyproject.toml` now too https://coverage.readthedocs.io/en/7.5.4/config.html#sample-file (although let's migrate to that in a different PR)

File: _build_meta.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+"""
+Comply with PEP 517's restictions on in-tree backends.
```
Comment (glyph) [inline _build_meta.py:2]: Love the explanation, thanks.

File: src/incremental/__init__.py [hunk 3]
```diff
@@ -360,36 +348,173 @@ def getVersionString(version):  # type: (Version) -> str
...
+
+        - src/{package}
+        - {package}
+
+    Where {package} is downcased.
+    """
+    src_dir = os.path.join(path, "src", package.lower())
+    current_dir = os.path.join(path, package.lower())
+
+    if os.path.isdir(src_dir):
```
Comment (glyph) [inline src/incremental/__init__.py:365]: (ugh at some point let's at least use pathlib here)

File: src/incremental/__init__.py [hunk 4]
```diff
@@ -360,36 +348,173 @@ def getVersionString(version):  # type: (Version) -> str
...
+        import tomli as tomllib
 
+    return tomllib.load(f)
 
-from ._version import __version__  # noqa: E402
+
+@dataclass(frozen=True)
+class _IncrementalConfig:
+    """
+    @ivar package: The package name, capitalized as in the package metadata.
```
Comment (glyph) [inline src/incremental/__init__.py:451]: FYI, you can just put these docstrings inline underneath the attribute declaration annotation and pydoctor will pick 'em up

**Review by glyph — APPROVED** [review]
I love it, this is a huge feature upgrade. Thanks for doing this!

File: .coveragerc (hunk 1 above)
Comment (twm) [reply .coveragerc:10]: Yeah I figured there's waaaay too much random cleanup here already. (I'd also like to migrate to Ruff, etc.)

**Comment by twm** [comment]
Thank you for the swift review @glyph! I dropped in a missing newsfragment for #80 and will merge now.

**Pull request merged**